lightningd/plugin_hook.c

```diff
@@ -1,5 +1,35 @@
 #include <lightningd/plugin_hook.h>
 
+static struct plugin_hook *plugin_hook_by_name(const char *name)
+{
+	static struct plugin_hook **hooks = NULL;
+	static size_t num_hooks;
+	if (!hooks)
+		hooks = autodata_get(hooks, &num_hooks);
+
+	for (size_t i=0; i<num_hooks; i++)
+		if (streq(hooks[i]->name, name))
+			return hooks[i];
+	return NULL;
+}
+
+bool plugin_hook_register(struct plugin *plugin, const char *method)
+{
+	struct plugin_hook *hook = plugin_hook_by_name(method);
+	if (!hook) {
+		/* No such hook name registered */
+		return false;
+	} else if (hook->plugin != NULL) {
+		/* Another plugin already registered for this name */
```

Comment:
Yeah, that might become a problem once people start using a myriad of different plugins, but I see no clean to fan-out to multiple plugins (serial vs parallel), and more importantly how to do fan-in, i.e., given two or more results which one to pick (serial call, and first non-default is picked vs. some sort of priority vs. ...).

Further complicating this is that hooks might have different ways of handling this, i.e., a hook that is just used to synchronously save information somewhere is trivially safe to use in parallel and have multiple ones, while a hook that actually modifies the continuation is not.

So the solution I'm eyeing currently is that we just punt this sort of multiplexing to a multiplexer plugin, that will then run the child-plugins instead of registering them with `lightningd` directly, and that would then know how to handle things.

We should probably start out like this and if there is a need to, we can start moving this kind of functionality into the plugin_hooks as necessary (I personally don't think there will be a huge competition for hooks, since they cater really specific needs, and most users will be happy with the default implemented in `lightningd` anyway).

Reply:
Right I think this makes sense at the outset.  It seems worth noting somewhere doc wise to cut down on potential for confusion later on.